hw/ip/otbn/dv/rig/rig/gens/bad_branch.py

```diff
@@ -0,0 +1,136 @@
+                                   'the shape expected by the Branch generator.'
+                                   .format(insn.mnemonic))
+
+        self.imm_op_type = self.bne.operands[2].op_type
+
+        for insn in insns_file.insns:
+            if insn.mnemonic in ['beq', 'bne']:
+                weight = cfg.insn_weights.get(insn.mnemonic)
+                if weight > 0:
+                    self.weights.append(weight)
+                    self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Max/Min Offsets for BXX (-2048 * 2, 2047 * 2)
+        # We give 0 to get_op_val_range method because it tries to calculate
+        # range with regards to given PC.
+        imm_rng = self.imm_op_type.get_op_val_range(0)
+        assert imm_rng is not None
+
+        min_offset, max_offset = imm_rng
+
+        # Get known registers. We always have x0.
+        # So it should never fail.
+        known_regs = model.regs_with_known_vals('gpr')
+        assert known_regs is not None
+
+        # Pick a random register among known registers.
+        idx, value = random.choice(known_regs)
+
+        equals = []
+        not_equals = []
+
+        for reg_idx, reg_val in known_regs:
+            if reg_val == value:
+                equals.append(reg_idx)
+            else:
+                not_equals.append(reg_idx)
+
+        # Get the chosen base register index as grs1 and grs2 operand. This is
+        # because we want to branch to the faulty addresses with this snippet.
+        op_val_grs1 = idx
+        assert op_val_grs1 is not None
+
+        # We need to pick an out of bounds offset from all available values (A)
+        # Best way to solve this is exluce the set of valid choices (B)
+        # We know tgt_addr can be max : pc + max_offset (A_max)
+        # We know tgt_addr can be min : pc + min_offset (A_min)
+        # Aim: (PC + min_offset, PC + max_offset) - (0, imem_size)
+
+        # Choose target from (A_min, A_max-B_max)
+        local_max = model.pc + max_offset - program.imem_size
+        tgt_addr = random.randrange(model.pc + min_offset, local_max, 2)
+
+        # If chosen value is in B, push it out by adding B_max
+        if tgt_addr >= 0:
+            tgt_addr += program.imem_size
+        assert tgt_addr < 0 or tgt_addr > program.imem_size
+
+        off_enc = self.imm_op_type.op_val_to_enc_val(tgt_addr, model.pc)
+        assert off_enc is not None
+
+        # Pick a random instruction from the weighted list
+        chosen_insn = random.choices(self.insns, weights=self.weights)[0]
+
+        if chosen_insn.mnemonic == 'bne':
+            if not not_equals:
+                return None
```

Comment:
Since we always know that `equals` is nonempty, maybe the best thing is to make sure we pick `beq` when `not_equals` is empty? Something like this, maybe?
```python3
  if not_equals:
    chosen_insn = random.choices(self.insns, weights=self.weights)[0]
  else:
    chosen_insn = self.beq
    beq_weight = self.weights[0]
    if not beq_weight:
      return None

  grs2_choices = equals if chosen_insn.mnemonic == 'beq' else not_equals
  assert grs2_choices
  op_val_grs2 = random.choice(grs2_choices)
```

Reply:
Thanks! It definitely makes more sense that way.